### PR TITLE
sandbox/apparmor: don't let vendored apparmor conflict with system

### PR DESCRIPTION
**Pull request by alexmurray**
Don't enable the vendored apparmor if the system installed apparmor will try and load policy that would be generated by the vendored apparmor and hence may conflict with that by using newer features not supported by the system installed apparmor (LP: 2024637)

### CODE REVIEW

File: sandbox/apparmor/apparmor.go [hunk 1]
```diff
@@ -396,6 +398,30 @@ func probeParserFeatures() ([]string, error) {
 	return features, nil
 }
 
+func systemAppArmorLoadsSnapPolicy() bool {
+	// on older Ubuntu systems the system installed apparmor may try and
+	// load snapd generated apparmor policy (LP: #2024637)
+	f, err := os.Open(filepath.Join(dirs.GlobalRootDir, "/lib/apparmor/functions"))
+	if err != nil {
+		logger.Debugf("cannot open apparmor functions file: %v", err)
```
Comment (pedronis) [inline sandbox/apparmor/apparmor.go:406]: we probably shouldn't log if the error is the file not being there at all

**Review by pedronis — COMMENTED** [review]
lgtm, one comment

File: sandbox/apparmor/apparmor.go [hunk 2]
```diff
@@ -396,6 +398,30 @@ func probeParserFeatures() ([]string, error) {
...
+		return false
```
Comment (pedronis) [inline sandbox/apparmor/apparmor.go:407]: for clarity, this is optimistic, it's probably fine given the nature of the checks but wanted to point this out

File: sandbox/apparmor/apparmor.go (hunk 1 above)
Comment (mvo5) [reply sandbox/apparmor/apparmor.go:406]: I added the logging, I wonder if I should just remove it?

Comment (pedronis) [reply sandbox/apparmor/apparmor.go:406]: well, the question is a bit larger than logging, the question is also whether we should return true or false in the face of error. in principle we should return true not knowing, but then if the file is missing we might return false at least on != ubuntu

Comment (alexmurray) [reply sandbox/apparmor/apparmor.go:406]: this file is the one which contains the code to load snapd-generated apparmor profiles - and as far as I know it is only present on debian/ubuntu -  if we fail to open it then it either doesn't exist or something is really odd with permissions in which case it is likely not going to load any snapd apparmor profiles, so in this case I think returning false in all cases of error makes sense.

**Review by pedronis — APPROVED** [review]
thanks

**Review by mvo5 — APPROVED** [review]
Thank you, this looks good to me too

**Comment by mvo5** [comment]
Some of the spread tests fail with:
```
2023-06-23T07:55:38.1445298Z ##[error]2023-06-23 07:55:38 Error executing google:ubuntu-18.04-64:tests/main/snapd-snap:lxd (jun230724-637010) : 
...
2023-06-23T07:55:38.6576723Z + tests.cleanup defer sh -c /home/gopath/src/github.com/snapcore/snapd/tests/main/snapd-snap/snapd-cleanup.sh
2023-06-23T07:55:38.6576945Z + echo 'Ensure we restarted into the snapd snap'
2023-06-23T07:55:38.6577076Z Ensure we restarted into the snapd snap
2023-06-23T07:55:38.6577481Z + /home/gopath/src/github.com/snapcore/snapd/tests/lib/tools/journal-state match-log 'restarting into "/snap/snapd/'
2023-06-23T07:55:38.6577761Z + echo 'Ensure sandbox-features shows the internal apparmor_parser'
2023-06-23T07:55:38.6578027Z Ensure sandbox-features shows the internal apparmor_parser
2023-06-23T07:55:38.6578485Z + snap debug sandbox-features --required apparmor:parser:snapd-internal
2023-06-23T07:55:38.6578773Z error: sandbox feature not available: "apparmor:parser:snapd-internal"
2023-06-23T07:55:38.6578881Z -----
```
so spread tests also need an update

**Review by olivercalder — APPROVED** [review]
LGTM, thanks! Just one small possible typo.

File: tests/main/snapd-snap/task.yaml [hunk 3]
```diff
@@ -235,6 +235,12 @@ execute: |
     echo "Ensure we restarted into the snapd snap"
     "$TESTSTOOLS"/journal-state match-log  'restarting into "/snap/snapd/'
 
+    # see LP:2024637
+    if grep -q /var/lib/snapd/apparmor/ /lib/apparmor/functions; then
+       echo "SKIP: cannot test buildin apparmor parser until /lib/apparmor/functions stops loading the snapd profiles"
```
[the remark this answers is not in the document]
Comment (alexmurray) [reply tests/main/snapd-snap/task.yaml]: thanks - typo fixed in https://github.com/snapcore/snapd/pull/12909/commits/b68062ce62dd6d81431ca944aaadcd50f2d202bc

**Comment by mvo5** [comment]
The nested spread test `qemu-nested:ubuntu-18.04-64 .../tests/nested/manual/snapd-refresh-from-old` fails with:
```
$ remote.exec 'sudo snap install snapd-from-deb.snap --dangerous'
error: cannot perform the following tasks:
- Make snap "snapd" (unset) available to the system (systemctl command [start snapd.apparmor.service] failed with exit status 1: Job for snapd.apparmor.service failed because the control process exited with error code.
See "systemctl status snapd.apparmor.service" and "journalctl -xe" for details.
)
$ remote.exec sudo journalctl -u snapd.apparmor.service
...
Jun 28 07:17:22 localhost snapd-apparmor[1640]: AppArmor parser error for /var/lib/snapd/apparmor/profiles/snap-confine.snapd.19626 in /var/lib/snapd/apparmor/snap-confine/cap-bpf at line 2: Invalid capability bpf.
Jun 28 07:17:22 localhost snapd-apparmor[1640]: AppArmor parser error for /var/lib/snapd/apparmor/profiles/snap-confine.snapd.x1 in /var/lib/snapd/apparmor/snap-confine/cap-bpf at line 2: Invalid capability bpf.
Jun 28 07:17:22 localhost snapd-apparmor[1640]: AppArmor parser error for /var/lib/snapd/apparmor/profiles/snap.pc.hook.configure in /var/lib/snapd/apparmor/profiles/snap.pc.hook.configure at line 6: Could not open 'if'
```

So it looks like that with the new snapd from edge that uses the vendored apparmor we cannot go back to this version of apparmor because (presumably) the new snapd code that no longer calls the vendored apparmor runs to start the profile.

**Comment by mvo5** [comment]
> The nested spread test `qemu-nested:ubuntu-18.04-64 .../tests/nested/manual/snapd-refresh-from-old` fails with:
> 
> ```
> $ remote.exec 'sudo snap install snapd-from-deb.snap --dangerous'
> error: cannot perform the following tasks:
> - Make snap "snapd" (unset) available to the system (systemctl command [start snapd.apparmor.service] failed with exit status 1: Job for snapd.apparmor.service failed because the control process exited with error code.
> See "systemctl status snapd.apparmor.service" and "journalctl -xe" for details.
> )
> $ remote.exec sudo journalctl -u snapd.apparmor.service
> ...
> Jun 28 07:17:22 localhost snapd-apparmor[1640]: AppArmor parser error for /var/lib/snapd/apparmor/profiles/snap-confine.snapd.19626 in /var/lib/snapd/apparmor/snap-confine/cap-bpf at line 2: Invalid capability bpf.
> Jun 28 07:17:22 localhost snapd-apparmor[1640]: AppArmor parser error for /var/lib/snapd/apparmor/profiles/snap-confine.snapd.x1 in /var/lib/snapd/apparmor/snap-confine/cap-bpf at line 2: Invalid capability bpf.
> Jun 28 07:17:22 localhost snapd-apparmor[1640]: AppArmor parser error for /var/lib/snapd/apparmor/profiles/snap.pc.hook.configure in /var/lib/snapd/apparmor/profiles/snap.pc.hook.configure at line 6: Could not open 'if'
> ```
> 
> So it looks like that with the new snapd from edge that uses the vendored apparmor we cannot go back to this version of apparmor because (presumably) the new snapd code that no longer calls the vendored apparmor runs to start the profile.

Looking at this a bit more we have code in `handlers.go` maybeRemoveAppArmorProfilesOnSnapdDowngrade() but it only looks at the snapd version and assumes that only snapd version "downgrade" can result in not using the internal parser anymore. Which is not the case here.

**Comment by pedronis** [comment]
> Looking at this a bit more we have code in handlers.go maybeRemoveAppArmorProfilesOnSnapdDowngrade() but it only looks at the snapd version and assumes that only snapd version "downgrade" can result in not using the internal parser anymore. Which is not the case here.

you are right, that means it needs to look at the services too, which also means that the logic based on that needs also to be run on refreshes/reverts of the boot base?

**Comment by mvo5** [comment]
I looked a bit more and I think the issue is the following:
1. snapd from edge/beta is installed on UC18
2. The install of this works fine, internal apparmor parer is used and profiles are written accordingly
3. But on each boot "apparmor.service" will now fail
4. When the fixed snapd from this PR is installed the *existing* snapd will rewrite security profiles and this will use the internal AppArmorParser because the the *broken* snapd is still running
5. But the *existing* snapd have rewritten /etc/systemd/system/snapd.apparmor.service (in LinkSnap) to point to the *new* revision from this PR
6. so new snapd-apparmor is run and now does *not* use the vendored apparmor because it detects that the system is not capable of that
7a. which means we could add a workaround into snapd-apparmor 
7b. *or* we leave it broken for now and it will fix itself automatically once 18.04/core18 ships with the SRU where we stop loading from /var/lib/snapd/apparmor/profiles in the systems "apparmor.service"

I'm inclined to go with option (7b) because only users on UC18/18.04 who run edge or beta are affected by the bug

**Comment by mvo5** [comment]
And indeed with the following simulation of a fixed `apparmor.service` in core18 the test passes:
```
diff --git a/tests/nested/manual/snapd-refresh-from-old/task.yaml b/tests/nested/manual/snapd-refresh-from-old/task.yaml
index 5c9d1082a5..26983f8e02 100644
--- a/tests/nested/manual/snapd-refresh-from-old/task.yaml
+++ b/tests/nested/manual/snapd-refresh-from-old/task.yaml
@@ -46,6 +46,11 @@ execute: |
     remote.exec "retry --wait 1 -n 10 sh -c 'snap changes | MATCH \".* Done .* Refresh snaps.*"core18"\"'"
   fi
 
+  # workaround bug in core18
+  sed 's#PROFILES_VAR="/var/lib/snapd/apparmor/profiles"#PROFILE_VAR="/non-existing"#g' /lib/apparmor/functions > /tmp/functions.fixed
+  remote.push /tmp/functions.fixed /tmp/functions.fixed
+  remote.exec sudo mount -o bind /tmp/functions.fixed /lib/apparmor/functions
+
   echo "Now refresh snapd with current tree"
   remote.push "snapd-from-deb.snap"
   remote.exec "sudo snap install snapd-from-deb.snap --dangerous" || true
```

**Pull request merged**